opm/simulators/linalg/bda/rocsparseWellContributions.cpp

```diff
@@ -0,0 +1,239 @@
+/*
+  Copyright 2020 Equinor ASA
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <config.h> // CMake
+
+// MultisegmentWellContribution includes the cuda runtime if found by CMake
+// this leads to inclusion of both amd_hip_vector_types.h and vector_types.h
+// which both define vector types like uchar2, short3 and double4.
+// Restore the value (if defined) afterwards.
+#ifdef HAVE_CUDA
+#define HIP_HAVE_CUDA_DEFINED HAVE_CUDA
+#endif
+
+#undef HAVE_CUDA
+
+#include <opm/simulators/linalg/bda/rocsparseWellContributions.hpp>
+
+#ifdef HIP_HAVE_CUDA_DEFINED
+#define HAVE_CUDA HIP_HAVE_CUDA_DEFINED
+#undef HIP_HAVE_CUDA_DEFINED
+#endif
+
+#include <opm/common/OpmLog/OpmLog.hpp>
+#include <opm/common/ErrorMacros.hpp>
+
+
+#include <hip/hip_runtime.h>
+
+
+#define HIP_CHECK(stat)                               \
+    {                                                 \
+        if(stat != hipSuccess)                        \
+        {                                             \
+            OPM_THROW(std::logic_error, "HIP error"); \
+        }                                             \
+    }
+
+namespace Opm
+{
+
+#ifdef __HIP__
+/// HIP kernel to apply the standard wellcontributions
+__global__ void stdwell_apply(
+            const double *Cnnzs,
+            const double *Dnnzs,
+            const double *Bnnzs,
+            const unsigned *Ccols,
+            const unsigned *Bcols,
+            const double *x,
+            double *y,
+            const unsigned dim,
+            const unsigned dim_wells,
+            const unsigned *val_pointers)
+{
+    unsigned wgId = blockIdx.x;
```

Comment:
Please add a comment why this is a custom kernel and we do not reuse sparse marix-vector products, here.

Reply:
I am not sure why this is. Tong Dong, maybe you can comment on this?

Reply:
This is simply a copy of the OpenCL custom kernel, which probably existed either to limit the dependencies, or because no easy library was available.
Given that rocsparse is already a dependency, it makes sense to reuse its spmv kernel.
However, in this kernel every well is applied in parallel.
To do that in a standard spmv kernel, all wells of a B, C or D matrix should probably be superimposed/merged into 1 matrix, such that no nnz of a well is in the same row/column as a nnz of another well.
I also remember trying to store the result of B*D^-1*C^T and using that to apply the wellcontributions, but that didn't work at the time.